planetiler-core/src/main/java/com/onthegomap/planetiler/util/LoopLineMerger.java

```diff
@@ -0,0 +1,438 @@
+        return candidate.cost;
+      }
+
+      for (var edge : current.getEdges()) {
+        var neighbor = edge.to;
+        if (neighbor != exclude) {
+          double newDist = candidate.cost + edge.length;
+          double prev = bestDistance.getOrDefault(neighbor.id, Double.POSITIVE_INFINITY);
+          if (newDist < prev) {
+            bestDistance.put(neighbor.id, newDist);
+            double heuristic = newDist + neighbor.distance(end);
+            if (heuristic <= maxLength) {
+              frontier.offer(new Candidate(neighbor, newDist, heuristic));
+            }
+          }
+        }
+      }
+    }
+    return Double.POSITIVE_INFINITY;
+  }
+
+  private void removeShortStubEdges(double stubMinLength) {
+    for (var node : output) {
+      for (var edge : List.copyOf(node.getEdges())) {
+        if (edge.length < stubMinLength &&
+          (edge.from.getEdges().size() == 1 || edge.to.getEdges().size() == 1 || edge.from == edge.to)) {
+          edge.remove();
+        }
+      }
+    }
+  }
+
+  private void removeShortEdges() {
+    for (var node : output) {
+      for (var edge : List.copyOf(node.getEdges())) {
+        if (edge.length < minLength) {
+          edge.remove();
+        }
+      }
+    }
+  }
+
+  public List<LineString> getMergedLineStrings() {
+    List<List<Coordinate>> edges = nodeLines(input);
+    buildNodes(edges);
+
+    merge();
+
+    if (loopMinLength > 0.0) {
+      breakLoops();
+      merge();
+    }
+
+    if (minLength > 0.0) {
+      double step = 1.0 / precisionModel.getScale();
+      for (double stubMinLength = 0.0; stubMinLength < minLength; stubMinLength += step) {
+        removeShortStubEdges(stubMinLength);
+        merge();
+      }
+      removeShortStubEdges(minLength);
+      merge();
+      // minLength = 10 * 0.0625;
+      // removeShortEdges();
+      // merge();
+    }
+
+    List<LineString> result = new ArrayList<>();
+
+    for (var node : output) {
+      for (var edge : node.getEdges()) {
+        if (edge.main) {
+          result.add(factory.createLineString(edge.coordinates.toArray(Coordinate[]::new)));
+        }
+      }
+    }
+
+    return result;
+  }
+
+  public List<LineString> getMergedByAngle() {
+    List<List<Coordinate>> edges = nodeLines(input);
+    buildNodes(edges);
+
+    mergeByAngle();
+
+    List<LineString> result = new ArrayList<>();
+
+    for (var node : output) {
+      for (var edge : node.getEdges()) {
+        if (edge.main) {
+          result.add(factory.createLineString(edge.coordinates.toArray(Coordinate[]::new)));
+        }
+      }
+    }
+
+    return result;
+  }
+
+  private double length(List<Coordinate> edge) {
+    Coordinate last = null;
+    double length = 0;
+    // we only care about the length of lines if they are < the limit
+    // so stop counting once we exceed it
+    double maxLengthToTrack = Math.max(minLength, loopMinLength);
+    if (maxLengthToTrack <= 0) {
+      return Double.POSITIVE_INFINITY;
+    }
```

Comment:
this was an optimization, when things settle a bit more you can try the benchmarks with and without the short-circuiting and if it doesn't make any difference than just remove it so this just computes the full length of the segment.